File: zig/circular_buffer/src/circular_buffer.zig
const std = @import("std");

pub const BUFFER_SIZE = 256;

pub const CircularBufferError = error{
    BufferEmpty,
    BufferFull,
};

pub const CircularBuffer = struct {
    buffer: [BUFFER_SIZE]u8,
    head: usize,
    tail: usize,

    pub fn new() CircularBuffer {
        return CircularBuffer{
            .buffer = [_]u8{0} ** BUFFER_SIZE,
            .head = 0,
            .tail = 0,
        };
    }

    pub fn is_full(self: *const CircularBuffer) bool {
        return ((self.head + 1) % BUFFER_SIZE) == self.tail;
    }

    pub fn is_empty(self: *const CircularBuffer) bool {
        return self.head == self.tail;
    }

    pub fn write(self: *CircularBuffer, data: u8) !void {
        if (self.is_full()) {
            return CircularBufferError.BufferFull;
        }

        self.buffer[self.head] = data;
        self.head = (self.head + 1) % BUFFER_SIZE;
    }

    pub fn read(self: *CircularBuffer) !u8 {
        if (self.is_empty()) {
            return CircularBufferError.BufferEmpty;
        }

        const data = self.buffer[self.tail];
        self.tail = (self.tail + 1) % BUFFER_SIZE;

        return data;
    }
};

File: go/circular_buffer/circular_buffer.go
package main

import (
	"errors"
)

const BUFFER_SIZE = 256

type CircularBuffer struct {
	buffer [BUFFER_SIZE]byte
	head   int
	tail   int
}

var (
	ErrBufferEmpty = errors.New("buffer is empty")
	ErrBufferFull  = errors.New("buffer is full")
)

func NewCircularBuffer() *CircularBuffer {
	return &CircularBuffer{
		head: 0,
		tail: 0,
	}
}

func (cb *CircularBuffer) IsFull() bool {
	return ((cb.head + 1) % BUFFER_SIZE) == cb.tail
}

func (cb *CircularBuffer) IsEmpty() bool {
	return cb.head == cb.tail
}

func (cb *CircularBuffer) Write(data byte) error {
	if cb.IsFull() {
		return ErrBufferFull
	}

	cb.buffer[cb.head] = data
	cb.head = (cb.head + 1) % BUFFER_SIZE

	return nil
}

func (cb *CircularBuffer) Read() (byte, error) {
	if cb.IsEmpty() {
		return 0, ErrBufferEmpty
	}

	data := cb.buffer[cb.tail]
	cb.tail = (cb.tail + 1) % BUFFER_SIZE

	return data, nil
}
